### PR TITLE
decomp: add docstring support to relevant places in all-types

### PR DESCRIPTION
**Pull request by xTVaser**
This will hopefully help us with better documenting the code as we go.  You can now put docstrings in most relevant places in the `all-types` file and they will carry through to the decompiled output.

This also adds support on the compiler side to atleast safely parse the docstrings, they aren't used yet as there isn't much to use it for other than lsp integration / maybe better error messages / repl hints.  A later project!

Some examples:

```lisp
(deftype vec4s (uint128)
  "hello world
   type docstring for vec4s"
  ((x float :offset 0)
   (y float :offset 32)
   (z float :offset 64)
   (w float :offset 96))
  :method-count-assert 9
  :size-assert         #x10
  :flag-assert         #x900000010
  )
```

```lisp
(define-extern identity "hello world identity" (function object object))
```

```lisp
(deftype bounding-box (structure)
  ((min vector :inline :offset-assert 0)
   (max vector :inline :offset-assert 16)
   )
  :method-count-assert 21
  :size-assert         #x20
  :flag-assert         #x1500000020
  (:methods
    (add-spheres! "method docstring!" (_type_ (inline-array sphere) int) int 9)
    (add-box! (_type_ bounding-box) int 10)
```

### CODE REVIEW

**Comment by ManDude** [comment]
Is it possible to move the docstring to the end of the form in `define-extern` and the method declarations? I'd prefer to have all that relevant information before the optional docstring.

**Comment by xTVaser** [comment]
It's possible but I'd rather not.  In regards to `define-extern` I'd rather it follow the same order as it would for the actual code, which looking at our code / other lisps the convention is generally `name > docstring > definition`

By extension, I think it's most intuitive for it to be `method name > docstring > definition` but I'm less opinionated since this doesn't match 1-1 with the output code.

**Comment by ManDude** [comment]
The convention for functions is `name > arguments > docstring > code` though.

**Comment by xTVaser** [comment]
Method definitions in the `deftype` have no `code` portion, so in my personal preference / what makes intuitive sense to me is the arg list there is the definition part.

If we are going to put everything after an arg list then the docstring in a `deftype` should be changed from

``lisp
(deftype vec4s (uint128)
  "hello world
   type docstring for vec4s"
  ((x float :offset 0)
   (y float :offset 32)
   (z float :offset 64)
   (w float :offset 96))
```

to

```lisp
(deftype vec4s (uint128)
  ((x float :offset 0)
   (y float :offset 32)
   (z float :offset 64)
   (w float :offset 96))
  "hello world
   type docstring for vec4s"
...
```

Which I would also be against.

At the end of the day this is bike-shedding over personal preference, I don't care to spend much time debating it.

**Comment by ManDude** [comment]
You could just make it some sort of named argument instead of forcing a specific (and potentially confusing to users?) order. Like how you can put `:methods` anywhere on the type.

**Comment by xTVaser** [comment]
It would be the only place that the docstring has to be a named argument though....

**Comment by ManDude** [comment]
Doesn't seem that bad to me. It's to be the standard way we've added things to a `deftype`.

**Comment by xTVaser** [comment]
I disagree, I'll wait for more of a consensus on the direction we want to go then, as this back and forth isn't productive.

File: common/type_system/TypeSystem.cpp [hunk 1]
```diff
@@ -1707,7 +1714,11 @@ std::string TypeSystem::generate_deftype_footer(const Type* type) const {
   std::string methods_string;
   auto new_info = type->get_new_method_defined_for_type();
   if (new_info) {
-    methods_string.append("(new (");
+    methods_string.append("(new ");
+    if (new_info->docstring) {
+      methods_string.append(fmt::format("\"{}\" ", new_info->docstring.value()));
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply common/type_system/TypeSystem.cpp:1719]: ah i forgot to revert this one, yeah the `new` method is the only builtin one you can override i believe? or atleast in practice the only one that ends up actually being overridden.

File: decompiler/util/DecompilerTypeSystem.cpp [hunk 2]
```diff
@@ -161,16 +170,16 @@ bool DecompilerTypeSystem::lookup_flags(const std::string& type, u64* dest) cons
...
+void DecompilerTypeSystem::add_symbol(const std::string& name,
+                                      const TypeSpec& type_spec,
+                                      const SymbolMetadata& symbol_metadata) {
   add_symbol(name);
   auto skv = symbol_types.find(name);
   if (skv == symbol_types.end() || skv->second == type_spec) {
     symbol_types[name] = type_spec;
-    if (definition_info) {
-      symbol_definition_info[name] = definition_info.value();
+    // TODO - could get rid of this if there is a way to go from TypeSpec -> full Type
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply decompiler/util/DecompilerTypeSystem.cpp:180]: The benefit (if i recall correctly) is that i wouldnt need a separate map, i could just lookup the `Type` which has additional information which we don't want to associate with the minimal `TypeSpec` type

File: decompiler/IR2/Form.cpp [hunk 3]
```diff
@@ -464,22 +464,25 @@ goos::Object SetFormFormElement::to_form_internal(const Env& env) const {
   }
 }
 
-goos::Object SetFormFormElement::to_form_for_define(const Env& env) const {
+goos::Object SetFormFormElement::to_form_for_define(
+    const Env& env,
+    const std::optional<std::string>& docstring) const {
+  std::vector<goos::Object> forms = {pretty_print::to_symbol("define"), m_dst->to_form(env)};
+  if (docstring) {
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply decompiler/IR2/Form.cpp:471]: How would this be parsed in a consistent/sane way?  Doesn't the reader ignore comments?

**Pull request merged**